module4-authentication-and-security/r1.1.1-authentication-persistence/README.md

```diff
@@ -0,0 +1,268 @@
+
+### PRO: Flexibility and Performance
+
+Flexibility and enhanced overall performance are other important aspects when it comes to token-based authentication as they can be used across multiple servers and they can offer authentication for diverse websites and applications at once.
+
+This helps in encouraging more collaboration opportunities between enterprises and platforms for a flawless experience.
+
+### PRO: Tokens Offer Robust Security
+
+Since tokens like JWT are stateless, only a secret key can validate it when received at a server-side application, which was used to create it.
+
+Hence they’re considered the best and the most secure way of offering authentication.
+
+Tokens act as a storage for the user’s credentials and when the token travels between the server or the web browser, the stored credentials are never compromised.
+
+### CON: Compromised Secret Key
+
+The best and the worst thing about JWT is that it relies on just one Key.
+
+Consider that this Key could be leaked by a careless or rogue developer/administrator, and then the whole system is compromised! The attacker (who has access to the Key) can easily access all user data if he has the user-id which can easily be acquired.
+
+The only way to recover from this point is to generate a new Key (Key-pair) that will be used across systems from here on. This would mean that all the existing client tokens are invalidated, and every user would have to login again. Imagine if one day all Facebook users are suddenly logged out.
+
+### CON: JWTs expire at specific intervals
+
+When a JWT is created it is given a specific expiration instant. The life of a JWT is definitive and it is recommended that it is somewhat small (think minutes not hours).
+
+Compared with traditional sessions, JWTs are quite different. Sessions are always a specific duration from the last interaction with the user. This means that if the user clicks a button, their session is extended. If you think about most applications you use, this is pretty common. You are logged out of the application after a specific amount of inactivity.
+
+JWTs, on the other hand, are not extended on user interaction. Instead, they are programmatically replaced by creating a new JWT for the user.
+
+To solve this problem, most applications use refresh tokens. Refresh tokens are opaque tokens that are used to generate new JWTs. Refresh tokens also need to expire at some point, but they can be more flexible in this mechanism because they are persisted in the identity provider or database.
+
+### CON: Data Overhead
+
+The overall size of a JWT is quite more than that of a normal session token, which makes it longer whenever more data is added to it.
+
+So, if you’re adding more claims in the token, it will impact the overall loading speed and thus hamper the user experience.
```

Comment:
the JWT is going to be so trivial in size compared to most webpage requests that i would consider this encouraging students to think about premature optimizations